CIP-plutus-modules/README.md

```diff
@@ -0,0 +1,1599 @@
+significant demand for modules in the context of smart contracts, even
+if the total contract code still remains relatively small.
+
+
+## Specification
+
+### Adding modules to UPLC
+
+This CIP provides the simplest possible way to split scripts across
+multiple UTxOs; essentially, it allows any closed subterm to be
+replaced by its hash, whereupon the term can be supplied either as a
+witness in the invoking transaction, or via a [reference script](https://cips.cardano.org/cip/CIP-0033) in that
+transaction. To avoid any change to the syntax of UPLC, hashes are
+allowed only at the top-level (so to replace a deeply nested subterm
+by its hash, we need to first lambda-abstract it). This also places
+all references to external terms in one place, where they can easily
+be found and resolved. Thus we need only change the definition of a
+`Script`; instead of simply some code, it becomes the application of
+code to zero or more arguments, given by hashes.
+
+Currently, the definition of “script” used by the ledger is (approximately):
+```
+newtype Script = Script ShortByteString
+```
+We change this to:
+```
+newtype CompleteScript = CompleteScript ShortByteString
+
+newtype Arg = ScriptArg ScriptHash
+
+data Script =
+  ScriptWithArgs { head :: CompleteScript, args :: [Arg] }
+
+-- hash of a Script, not a CompleteScript
+type ScriptHash = ByteString
+```
+
+Scripts in transactions, and on the chain, are represented in this
+way, with dependencies that must be supplied in a transaction using
+the script. During phase 2 verification we need to resolve the
+arguments of each script before running it:
+```
+resolveScriptDependencies
+  :: Map ScriptHash Script
+  -> Script
+  -> Maybe CompleteScript
+resolveScriptDependencies preimages = go
+  where
+    go (ScriptWithArgs head args) = do
+      argScripts <- traverse lookupArg args
+      pure $ applyScript head argScripts
+      where
+        lookupArg :: Arg -> Maybe CompleteScript
+        lookupArg (ScriptArg hash) = do
+          script <- lookup hash preimages
+          go script
+```
+The `preimages` map is the usual witness map constructed by the ledger,
+so in order for a script hash argument to be resolved, the transaction
+must provide the pre-image in the usual way. Note that arguments are
+mapped to a `Script`, not a `CompleteScript`, so the result of looking
+up a hash may contain further dependencies, which need to be resolved
+recursively. A transaction must provide witnesses for *all* the
+recursive dependencies of the scripts it invokes.
+
+The only scripts that can be run are complete scripts, so the type of
+`runScript` changes to take a `CompleteScript` instead of a `Script`.
+
+#### Variation: Lazy Loading
+
+With this design, if any script hash is missing from the `preimages`,
+then the entire resolution fails. As an alternative, we might replace
+missing subterms by a dummy value, such as `builtin unit`, thus:

```

Comment:
You can't use `error` because UPLC is strict, so module values will tend to get evaluated eagerly, triggering the error. We could use `error` if module values were all delayed, for example. The nice thing about `()` is that either it doesn't work (which is fine), or it works... but in that case it just means you've accidentally supplied the right value!

Reply:
> You can't use `error` because UPLC is strict

Doh.  In my defence, it was late at night.  I think there is still an issue (mentioned later) that all terms of type `()` aren't equaivalent because of `error`.  It's hard to imagine that that would arise in reality though.